crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -176,3 +243,102 @@ mod observe {
+            };
+            Ok(Self {
+                client,
+                base_url,
+                api_key,
+                chain,
+                quote_token,
+            })
+        }
+    }
+
+    // It is ok to call this API without an API for local testing purposes as it is
+    // difficulty to hit the rate limit manually
+    const BASE_API_URL: &str = "https://api.coingecko.com/api/v3/simple/token_price";
+
+    // We also need to test the PRO API, because batch requests aren't available in
+    // the free version
+    const BASE_API_PRO_URL: &str = "https://pro-api.coingecko.com/api/v3/simple/token_price";
+
+    #[tokio::test]
+    #[ignore]
+    async fn works() {
+        let native_token = addr!("C02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2");
+        let instance = CoinGecko::new_for_test(
+            Client::default(),
+            Url::parse(BASE_API_URL).unwrap(),
+            None,
+            1,
+        )
+        .unwrap();
+
+        let estimated_price = instance.estimate_native_price(native_token).await.unwrap();
+        // Since the WETH precise price against ETH is not always exact to 1.0 (it can
+        // vary slightly)
+        assert!((0.95..=1.05).contains(&estimated_price));
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn works_xdai() {
+        // USDT
+        let native_token = addr!("4ECaBa5870353805a9F068101A40E0f32ed605C6");
+        let instance = CoinGecko::new_for_test(
+            Client::default(),
+            Url::parse(BASE_API_PRO_URL).unwrap(),
+            env::var("COIN_GECKO_API_KEY").ok(),
+            100,
+        )
+        .unwrap();
+
+        let estimated_price = instance.estimate_native_price(native_token).await.unwrap();
+        // Since the USDT precise price against XDAI is not always exact to 1.0
+        // (it can vary slightly)
+        assert!((0.95..=1.05).contains(&estimated_price));
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn works_multiple_tokens() {
+        let usdt_token = addr!("4ECaBa5870353805a9F068101A40E0f32ed605C6");
+        let usdc_token = addr!("2a22f9c3b484c3629090FeED35F17Ff8F88f76F0");
+        let instance = CoinGecko::new_for_test(
+            Client::default(),
+            Url::parse(BASE_API_PRO_URL).unwrap(),
+            env::var("COIN_GECKO_API_KEY").ok(),
+            100,
+        )
+        .unwrap();
+
+        let estimated_price = instance
+            .fetch_native_prices(&HashSet::from([usdt_token, usdc_token]))
+            .await
+            .unwrap();
+        let usdt_price = estimated_price.get(&usdt_token).unwrap().clone();
+        let usdc_price = estimated_price.get(&usdc_token).unwrap().clone();
+        // Since the USDT/USDC precise price against XDAI is not always exact to
+        // 1.0 (it can vary slightly)
+        assert!((0.95..=1.05).contains(&usdt_price.unwrap()));
+        assert!((0.95..=1.05).contains(&usdc_price.unwrap()));
+    }
```

Comment:
worth adding a test to check the bulk limit too

Reply:
Not sure about that since this is nothing we can prevent given the current setup anyway.
To reasonably enforce that in a way the can be tested the `BufferedPriceFetching` would have to spawn its own background task with the enforced limit.
At the moment this gets set up outside of this component so even if we have a test that detects some sort of error if `batch_size > 20` it would not help us prevent this error in the actual business logic.